### ISSUE TITLE
Non-conforming refinement

### ISSUE THREAD

**Issue by junkudo** [mfem/mfem#739]
Hey,
I'm having some issues in my code that I believe is related to some inconsistencies in the face neighbor data when I perform refinement on a non-conforming mesh.

If I look at the sizes at some of the face neighbor data structures formed during `mfem::ParMesh::ExchangeFaceNbrData()`, the sizes of the data different is different if I...
1.) Exchange data only after I refine
```c++
#include <iostream>
#include "mpi.h"
#include "mfem.hpp"

// 1.) Exchange data only after I refine
int main(int argc, char* argv[]) {
   MPI_Init(&argc, &argv);
   MPI_Comm comm = MPI_COMM_WORLD;
   int myrank;
   MPI_Comm_rank(comm, &myrank);

   // Create non-conforming mesh
   mfem::Mesh mesh(10, 10, mfem::Element::QUADRILATERAL, 1, 1.0, 1.0);
   mesh.EnsureNCMesh();
   mfem::ParMesh pmesh(comm, mesh);

   // Perform refinement through GeneralRefinement
   pmesh.UniformRefinement();

   pmesh.ExchangeFaceNbrData();
   std::cout << myrank << ": After refinement - face nbr elements size: " << pmesh.face_nbr_elements.Size() << std::endl;
}
```
gives me
```bash
0: After refinement - face nbr elements size: 20
1: After refinement - face nbr elements size: 20
```
---
2.) Exchange data before and after I refine
```c++
#include <iostream>
#include "mpi.h"
#include "mfem.hpp"

int main(int argc, char* argv[]) {
   MPI_Init(&argc, &argv);
   MPI_Comm comm = MPI_COMM_WORLD;
   int myrank;
   MPI_Comm_rank(comm, &myrank);

   // Create non-conforming mesh
   mfem::Mesh mesh(10, 10, mfem::Element::QUADRILATERAL, 1, 1.0, 1.0);
   mesh.EnsureNCMesh();
   mfem::ParMesh pmesh(comm, mesh);

   // Exchange face neighbor data
   pmesh.ExchangeFaceNbrData();
   std::cout << myrank << ": Before refinement - face nbr elements size: " << pmesh.face_nbr_elements.Size() << std::endl;

   // Perform refinement through GeneralRefinement
   pmesh.UniformRefinement();

   pmesh.ExchangeFaceNbrData();
   std::cout << myrank << ": After refinement - face nbr elements size: " << pmesh.face_nbr_elements.Size() << std::endl;
}
```
gives me
```bash
0: Before refinement - face nbr elements size: 10
0: After refinement - face nbr elements size: 10
1: Before refinement - face nbr elements size: 10
1: After refinement - face nbr elements size: 10
```

I think part of the problem is that the non-conforming refinement procedure as currently implemented does not seem to delete the face neighbor data (so the second `ExchangeFaceNbrData()` is a no-op).

Thanks for your help!
@kswartz14 @Dan2997925

**Comment by v-dobrev** [comment]
Hi @junkudo,

Can you try adding a call to `DeleteFaceNbrData()` (this is a method in `ParMesh`) in the beginning of `ParMesh::NonconformingRefinement()`?

Let us know if this resolves the issue, so we can make a branch/PR to fix it.

Thanks!

**Comment by junkudo** [comment]
Hey @v-dobrev,
Putting in `DeleteFaceNbrData()` seems to resolve the issue.
```bash
0: Before refinement - face nbr elements size: 10
1: Before refinement - face nbr elements size: 10
0: After refinement - face nbr elements size: 20
1: After refinement - face nbr elements size: 20
```
Thanks!

**Comment by v-dobrev** [comment]
OK, great! I will create a branch with the fix in a moment.

**Comment by junkudo** [comment]
Hey @v-dobrev, when can I expect this fix to be merged into the master branch?

**Comment by v-dobrev** [comment]
I just merged it.
